File: cucumber-expressions/javascript/src/group_builder.js
const Group = require('./group')

class GroupBuilder {
  constructor() {
    this._groupBuilders = []
    this._capturing = true
  }

  add(group_builder) {
    this._groupBuilders.push(group_builder)
  }

  build(match, nextGroupIndex) {
    const groupIndex = nextGroupIndex()
    const children = this._groupBuilders.map(gb =>
      gb.build(match, nextGroupIndex)
    )
    return new Group(
      match[groupIndex],
      match.index[groupIndex],
      match.index[groupIndex] + (match[groupIndex] || '').length,
      children
    )
  }

  setNonCapturing() {
    this._capturing = false
  }

  get capturing() {
    return this._capturing
  }

  get children() {
    return this._groupBuilders
  }

  moveChildrenTo(groupBuilder) {
    this._groupBuilders.forEach(child => groupBuilder.add(child))
  }
}

module.exports = GroupBuilder

File: cucumber-expressions/go/group_builder.go
package cucumberexpressions

type GroupBuilder struct {
	groupBuilders []*GroupBuilder
	capturing     bool
	source        string
}

func NewGroupBuilder() *GroupBuilder {
	return &GroupBuilder{
		capturing: true,
	}
}

func (g *GroupBuilder) Add(groupBuilder *GroupBuilder) {
	g.groupBuilders = append(g.groupBuilders, groupBuilder)
}

func (g *GroupBuilder) Build(submatches []*Submatch, indexIterator *IntIterator) *Group {
	submatch := submatches[indexIterator.Next()]
	children := make([]*Group, len(g.groupBuilders))
	for i, child := range g.groupBuilders {
		children[i] = child.Build(submatches, indexIterator)
	}
	return NewGroup(submatch.value, submatch.start, submatch.end, children)
}

func (g *GroupBuilder) SetNonCapturing() {
	g.capturing = false
}

func (g *GroupBuilder) Capturing() bool {
	return g.capturing
}

func (g *GroupBuilder) Children() []*GroupBuilder {
	return g.groupBuilders
}

func (g *GroupBuilder) MoveChildrenTo(groupBuilder *GroupBuilder) {
	for _, child := range g.groupBuilders {
		groupBuilder.Add(child)
	}
}

func (g *GroupBuilder) SetSource(value string) {
	g.source = value
}

func (g *GroupBuilder) Source() string {
	return g.source
}
